File: Prototype/Prototype.js
/**
 * Created by jamesyan on 2017/3/19.
 */

var shape=function(type,desc){
    this.type=type;
    this.draw=function(){
       console.log(desc)
    };
}

shape.prototype.getId=function(){
    return this.id;
}

shape.prototype.setId=function(id){
    this.id=id;
}
shape.prototype.getType=function(){
    return this.type;
}

shape.prototype.setType=function(type){
    this.type=type;
}

var Rectangle=function(){
    shape.call(this,"Rectangle","Inside Rectangle::draw() method");
}

Rectangle.prototype=Object.create(shape.prototype);

var Circle=function(){
    shape.call(this,"Circle","Inside Circle::draw() method");
}

Circle.prototype=Object.create(shape.prototype);

var Square=function(){
    shape.call(this,"Square","Inside Square::draw() method");
}

Square.prototype=Object.create(shape.prototype);


var ShapeCache=function(){
    this.shapeMap={};
}

ShapeCache.prototype.loadCache=function(){
    var circle=new Circle();
    circle.setId("1");
    this.shapeMap[circle.getId()]=circle;

    var square=new Square();
    square.setId("2");
    this.shapeMap[square.getId()]=square;

    var rectangle=new Rectangle();
    rectangle.setId("3");
    this.shapeMap[rectangle.getId()]=rectangle;

}

ShapeCache.prototype.getShape=function(shapeId){
    var shape=this.shapeMap[shapeId];
    return Object.create(shape);
}

File: Prototype/Prototype.go
package main

import (
   "fmt"
)

type InterShape interface {
   getId() int
   getType() string
   setId(int)
   setType()
   setTest(string)
   draw()
}

//父类
type Shape struct {
    id int
    Type string
}

func(s *Shape)setTest(name string){
   s.Type=name
}

func(s *Shape) getType() string{
   return s.Type
}

func(s *Shape) getId() int{
   return s.id
}

func(s *Shape)setId(id int){
   s.id=id
}

//子类
type Rectangle struct {
   Shape
}

func(s *Rectangle)setType(){
   s.Type="Rectangle"
}


func(s *Rectangle)draw(){
   fmt.Println("Inside Rectangle::draw() method");
}

type Circle struct {
   Shape
}

func(s *Circle)setType(){
   s.Type="Circle"
}


func(s *Circle)draw(){
   fmt.Println("Inside Circle::draw() method");
}

type Square struct {
   Shape
}

func(s *Square)setType(){
   s.Type="Square"
}


func(s *Square)draw(){
   fmt.Println("Inside Square::draw() method");
}

//创建shapeCache
type ShapeCache struct {
   shapeMap map[int]InterShape
}

func(s *ShapeCache)getShape(shapeId int) InterShape{
    shape:=s.shapeMap[shapeId]
    return shape//如何实现对象赋值？
}

func(s *ShapeCache)loadCache(){
   s.shapeMap=make(map[int]InterShape)
   circle:=new(Circle)
   circle.setId(1)
   circle.setType()
   s.shapeMap[circle.getId()]=circle

   square:=new(Square)
   square.setId(2)
   square.setType()
   s.shapeMap[square.getId()]=square

   rectangle:=new(Rectangle)
   rectangle.setId(3)
   rectangle.setType()
   s.shapeMap[rectangle.getId()]=rectangle
}


func main(){
   shapeCache:=new(ShapeCache)
   shapeCache.loadCache()
   clonedShape1:=shapeCache.getShape(1)
   fmt.Println("Shape1:",clonedShape1.getType())
   clonedShape1.draw()

   clonedShape2:=shapeCache.getShape(2)
   fmt.Println("Shape2:",clonedShape2.getType())
   clonedShape2.draw()

   clonedShape3:=shapeCache.getShape(3)
   fmt.Println("Shape3:",clonedShape3.getType())
   clonedShape3.draw()

   clonedShape4:=shapeCache.getShape(1)
   clonedShape4.setTest("Hello")
   fmt.Println("Shape4:",clonedShape4.getType())
   clonedShape4.draw()
   fmt.Println("Shape1:",clonedShape1.getType())
}


